database/user.go

```diff
@@ -224,26 +229,75 @@ func (db *DB) UserByStripeID(ctx context.Context, id string) (*User, error) {
 	return &u, nil
 }
 
+// UserBySub returns the user with the given sub. If `create` is `true` it will
+// create the user if it doesn't exist. The sub is the Kratos id of that user.
+func (db *DB) UserBySub(ctx context.Context, sub string, create bool) (*User, error) {
```

Comment:
Is there a reason why your getter methods come with a `create` bool? This feels pretty strange tbh. 
I'd expect a getter that returns a user and maybe a bool and error to indicates whether the user exists or not and a setter which creates a user and returns an error.

Reply:
The reason for that is that `accounts` wasn't responsible for user registration. So, the workflow was (still is):
- user registers with Kratos
- `webportal` maybe tries to call `accounts` in a post-registration hook and tell it about the new user but that might fail and we might end up with the user not in accounts. `webportal` cannot store that anywhere since it doesn't have a DB, so it cannot keep retrying the user registration.
- the user performs an action (upload, download, registry access, etc.) while logged in. Nginx calls `accounts` on the respective `track` endpoint and provides the user's JWT token which contains everything we need in order to register them. Note that nginx cannot try to call `track`, call a `create user` endpoint on error and then retry the `track` request - it can only issue a static set of API calls.
- `accounts` receives the `track` request, looks for the user in the DB and doesn't find it. It decides that it's a new user that it just hasn't been informed about and registers them.
- `accounts` uses a boolean flag in order to be able to use the same methods in situations in which the user we're trying to fetch shouldn't be auto-created.

Once we have removed Kratos and migrated all the data into accounts's MongoDB, we can remove that flag and stop creating users implicitly.

Reply:
F/U: #71